SITL: Add C++ library for JSON interface

TunaLobster opened the pull request:
We were working on creating a simulator from scratch to meet some specific needs and I realized there wasn't a plug-and-play way to add support for ArduPilot to a new sim. This is probably trivial for most, but I found this to be very useful for my work.

Right now this code is not the best and naming is open for suggestions. Ideally, there would be a proper way to change what pieces of JSON data to send to ArduPilot, but I am not sure if that should be done with API methods with default value NULL values and several statements similar to `if airseed is not None`. I know in Python I would use optional arguments, but I'm not sure what something similar would look like here.

TunaLobster commented:
Tested working on WSL 1 and 2 Ubuntu 20.04.

davidbuzz commented:
just curious, its still marked as wIP, but i think it should be mergable, yes..?  @TunaLobster ?

TunaLobster commented:
It works and gets the job done. It's not as complete as I would want it to be though. Testing with winsock and a better way to set the optional JSON interface values would be nice.

TunaLobster commented:
Added setters for windvane and airspeed. Screenshot is the whole thing running through WSL 1. Cleaned up some pieces too.

@davidbuzz I'll call it good for right now. Getting winsock and rangefinder is getting pushed down the road on my schedule.

peterbarker commented on libraries/SITL/examples/JSON/C++/Socket.h:
```diff
@@ -0,0 +1,76 @@
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  simple socket handling class for systems with BSD socket API
+ */
```
Remove

TunaLobster replied:
Fixed.

peterbarker commented on libraries/SITL/examples/JSON/C++/Socket.h:
```diff
@@ -0,0 +1,76 @@
+ */
+#pragma once
+
+// #include <AP_HAL/AP_HAL.h>
+// #if HAL_OS_SOCKETS
```
Remove

TunaLobster replied:
Fixed.

peterbarker commented on libraries/SITL/examples/JSON/C++/Socket.h:
```diff
@@ -0,0 +1,76 @@
+#include <arpa/inet.h>
+#include <sys/select.h>
+
+class SocketAPM {
```
Rename

Move to C++ code as being an internal thing?

TunaLobster replied:
Renamed to SocketExample

tridge commented on libraries/SITL/examples/JSON/C++/Socket.h:
```diff
@@ -0,0 +1,76 @@
+#include <arpa/inet.h>
+#include <sys/select.h>
+
+class SocketAPM {
```
can we please rename this to a different name than the AP_HAL/utility/Socket.h

TunaLobster replied:
Fixed.

tridge reviewed (changes requested):
add a README.md on how to build and test this

peterbarker commented on libraries/SITL/examples/JSON/C++/libAP_JSON.cpp:
```diff
@@ -0,0 +1,263 @@
+    // missed receives before declaring the FCS offline.
+
+    uint32_t waitMs;
+    if (arduPilotOnline)
+    {
```
```suggestion
    if (arduPilotOnline) {
```


... and generally ardupilot-ise

TunaLobster replied:
Fixed up the ones I recognize.

peterbarker reviewed (changes requested):
Short in-meeting review

TunaLobster commented:
A readme and a 1-D have been added. Fixed an issue where the rangefinder data is properly indexed at 1 rather than 0.

peterbarker commented:
Merged, thanks!
